Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIr.kt

```diff
@@ -0,0 +1,460 @@
+) : StubContainer {
+    override fun accept(visitor: StubIrVisitor) {
+        visitor.visitSimpleStubContainer(this)
+    }
+
+    override fun <T, R> accept(visitorAux: StubIrVisitorAux<T, R>, data: T): R {
+        return visitorAux.visitSimpleStubContainer(this, data)
+    }
+}
+
+val StubContainer.children: List<StubIrElement>
+    get() {
+        val classes: List<StubIrElement> = classes
+        val funcs: List<StubIrElement> = functions
+        return classes + properties + funcs + typealiases
+    }
+
+
+sealed class StubType {
+    abstract val nullable: Boolean
+}
+
+/**
+ * Marks that abstract value of such type can be passed as value.
+ */
+sealed class ValueStub
+
+class TypeParameterStub(
+        val name: String,
+        val upperBound: StubType? = null
+) {
+    fun getStubType(nullable: Boolean): StubType =
+            SymbolicStubType(name, nullable = nullable)
+
+}
+
+// Add variance if needed
+class TypeArgumentStub(val type: StubType)
+
+/**
+ * Wrapper over [KotlinType].
+ */
+class WrapperStubType(
+        val kotlinType: KotlinType
+) : StubType() {
+    override val nullable: Boolean
+        get() = when (kotlinType) {
+            is KotlinClassifierType -> kotlinType.nullable
+            is KotlinFunctionType -> kotlinType.nullable
+            else -> error("Unknown KotlinType: $kotlinType")
+        }
+}
+
+/**
+ * Wrapper over [Classifier].
+ */
+class ClassifierStubType(
+        val classifier: Classifier,
+        val typeArguments: List<TypeArgumentStub> = emptyList(),
+        override val nullable: Boolean = false
+) : StubType()
+
+
+/**
+ * Fallback variant for all cases where we cannot refer to specific [KotlinType] or [Classifier].
+ */
+class SymbolicStubType(
+        val name: String,
+        val typeArguments: List<TypeArgumentStub> = emptyList(),
+        override val nullable: Boolean = false
+) : StubType()
+
+/**
+ * Represents a source of StubIr element.
+ */
+sealed class StubOrigin {
+    /**
+     * Special case when element of IR was generated.
+     */
+    object None : StubOrigin()
+
+    class ObjCMethod(
+            val method: org.jetbrains.kotlin.native.interop.indexer.ObjCMethod,
+            val container: ObjCContainer
+    ) : StubOrigin()
+
+    class ObjCClass(
+            val clazz: org.jetbrains.kotlin.native.interop.indexer.ObjCClass
+    ) : StubOrigin()
+
+    class ObjCProtocol(
+            val protocol: org.jetbrains.kotlin.native.interop.indexer.ObjCProtocol
+    ) : StubOrigin()
+
+    class Enum(val enum: EnumDef) : StubOrigin()
+
+    class Function(val function: FunctionDecl) : StubOrigin()
+
+    class FunctionParameter(val parameter: Parameter) : StubOrigin()
+
+    class Struct(val struct: StructDecl) : StubOrigin()
+}
+
+interface StubElementWithOrigin : StubIrElement {
+    val origin: StubOrigin
+}
+
+interface AnnotationHolder {
+    val annotations: List<AnnotationStub>
+}
+
+sealed class AnnotationStub {
```

Comment:
Why not make annotation representation more low-level, e.g. class fq name + arguments?
I expect both output formats requiring annotations in this form.

Reply:
Since the list of annotation types is limited this approach is simpler from development point of view. 
It's easier to find usages of specific annotation type, easier to render (e.g. one can avoid rendering default parameter values) and it's less error-prone to typos and forgotten renamings that are quite common in low-level approach.